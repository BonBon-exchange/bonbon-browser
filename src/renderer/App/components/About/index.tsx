import { About } from './About';

export * from './About';
export default About;
