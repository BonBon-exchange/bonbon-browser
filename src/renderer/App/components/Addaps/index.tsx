import { Addaps } from './Addaps';

export * from './Addaps';
export default Addaps;
