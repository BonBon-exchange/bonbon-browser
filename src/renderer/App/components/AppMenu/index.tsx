import { AppMenu } from './AppMenu';

export * from './AppMenu';
export default AppMenu;
