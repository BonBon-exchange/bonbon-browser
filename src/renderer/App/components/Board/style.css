.Board__container {
  display: flex;
  height: 300vh;
  width: calc(100% - 50px);
  top: 0px;
  left: 50px;
  position: absolute;
  background: linear-gradient(
    200.96deg,
    #fedc2a -29.09%,
    #dd5789 51.77%,
    #7a2c9e 129.35%
  );
}
