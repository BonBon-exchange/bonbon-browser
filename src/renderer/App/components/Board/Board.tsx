/* eslint-disable promise/always-return */
/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable import/prefer-default-export */
/* eslint-disable no-use-before-define */
import clsx from 'clsx';
import { AnimatePresence } from 'framer-motion';
import React, { useCallback, useEffect, useState } from 'react';

import { Browser } from 'renderer/App/components/Browser';
import { BrowserProps } from 'renderer/App/components/Browser/Types';
import { useBoard } from 'renderer/App/hooks/useBoard';
import { useBrowserMethods } from 'renderer/App/hooks/useBrowserMethods';
import { useStoreHelpers } from 'renderer/App/hooks/useStoreHelpers';
import { useAppDispatch } from 'renderer/App/store/hooks';
import {
  setActiveBrowser,
  setBoardHeight,
} from 'renderer/App/store/reducers/Board';

import { BoardProps } from './Types';

import './style.scss';

export const Board: React.FC<BoardProps> = ({ isFullSize, boardId }) => {
  const board = useBoard();
  const dispatch = useAppDispatch();
  const { focus } = useBrowserMethods();
  const [items, setItems] = useState<BrowserProps[]>([]);
  const helpers = useStoreHelpers();

  const boardContainer = document.querySelector('#Board__container');

  const makeBrowsers = useCallback((sorted: BrowserProps[]) => {
    return sorted.map((b) => <Browser {...b} key={b.id} firstRendering />);
  }, []);

  const contextMenuListener = (e: MouseEvent) => {
    e.preventDefault();
    window.app.tools.showBoardContextMenu({ x: e.clientX, y: e.clientY });
  };

  useEffect(() => {
    if (!board.isFullSize) {
      setTimeout(() => {
        helpers.board
          .distributeWindowsByOrder(board.browsers)
          .then(() => {
            setTimeout(
              () => board.activeBrowser && focus(board.activeBrowser),
              300
            );
          })
          .catch(console.log);
      }, 300);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [board.isFullSize]);

  useEffect(() => {
    const toSort = [...board.browsers];
    const sorted = toSort.sort((a, b) => {
      return a.id > b.id ? 1 : -1;
    });
    setItems(sorted);

    if (boardContainer) {
      const max = board.browsers.reduce(
        (acc, val) => Math.max(acc, val.top + val.height + 100),
        0
      );

      // @ts-ignore
      boardContainer.style.height = `${Number(max)}px`;

      dispatch(setBoardHeight(max));
    }
  }, [board.browsers, dispatch, boardContainer]);

  useEffect(() => {
    window.app.board.setWindowsCount({
      boardId: board.id,
      count: board.browsers.length,
    });
  }, [board.browsers.length, board.id]);

  useEffect(() => {
    if (board.activeBrowser) {
      focus(board.activeBrowser, true);
      dispatch(setActiveBrowser(board.activeBrowser));
    }
  }, [board.activeBrowser, focus, dispatch]);

  // focus activeBrowser when componentDidMount
  useEffect(() => {
    setTimeout(() => {
      if (board.activeBrowser) focus(board.activeBrowser);
      if (boardId) {
        window.app.browser
          .getUrlToOpen()
          .then((res) => {
            if (res) helpers.browser.add({ url: res });
          })
          .catch(console.log);
      }
    }, 0);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    document
      .getElementById('Board__container')
      ?.addEventListener('contextmenu', contextMenuListener);

    return () =>
      document
        .getElementById('Board__container')
        ?.removeEventListener('contextmenu', contextMenuListener);
  }, []);

  return (
    <div
      id="Board__container"
      className={clsx({
        'Board__is-maximized': board.isFullSize || isFullSize,
        'Board__isnt-maximized': !board.isFullSize && !isFullSize,
      })}
    >
      <AnimatePresence>{makeBrowsers(items)}</AnimatePresence>
      <div className="Board__edge-snap-left" />
      <div className="Board__edge-snap-right" />
      <div className="Board__edge-snap-maximized" />
    </div>
  );
};
