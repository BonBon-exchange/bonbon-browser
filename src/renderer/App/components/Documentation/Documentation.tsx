/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable import/prefer-default-export */
import { useState, ReactElement, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import clsx from 'clsx';

import { CloseButton } from 'renderer/App/components/CloseButton';
import { KeyboardShortcuts } from './KeyboardShortcuts';
import { WebpagesWindows } from './WebpagesWindows';

import './style.scss';

import { DocumentationProps } from './Types';

export const Documentation: React.FC<DocumentationProps> = ({
  handleClose,
}: DocumentationProps) => {
  const { t } = useTranslation();
  const [selectedView, setSelectedView] = useState<ReactElement>(
    <KeyboardShortcuts />
  );
  const [scrollbarVisible, setScrollBarVisible] = useState<null | boolean>(
    false
  );

  const isScrollbarVisible = (element: HTMLElement | null) => {
    return element && element.scrollHeight > element.clientHeight;
  };

  useEffect(() => {
    setScrollBarVisible(
      isScrollbarVisible(document.getElementById('Documentation__right-panel'))
    );
  }, []);

  return (
    <div id="Documentation__container">
      <CloseButton
        handleClose={handleClose}
        customClass={clsx({
          'CloseButton__with-scrollbar': scrollbarVisible,
        })}
      />
      <div id="Documentation__left-panel">
        <h2>{t('Documentation')}</h2>
        <ul>
          <li onClick={() => setSelectedView(<KeyboardShortcuts />)}>
            {t('Keyboard shortcuts')}
          </li>
          <li
            onClick={() => setSelectedView(<WebpagesWindows />)}
            data-testid="documentation-webpages-link"
          >
            {t('Webpages windows')}
          </li>
        </ul>
      </div>
      <div id="Documentation__right-panel">{selectedView}</div>
    </div>
  );
};
