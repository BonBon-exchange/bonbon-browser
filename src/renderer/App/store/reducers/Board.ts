/* eslint-disable @typescript-eslint/no-unused-expressions */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 } from 'uuid';

import { WritableDraft } from 'immer/dist/internal';
import { BoardType } from 'renderer/App/components/Board/Types';
import { BrowserProps } from 'renderer/App/components/Browser/Types';

type UpdateBrowserType = {
  browserId: string;
  params: Partial<BrowserProps>;
};

type UpdateBrowserUrlType = {
  url: string;
  browserId: string;
};

type UpdateBrowserLoadingType = {
  isLoading: boolean;
  browserId: string;
};

type UpdateBrowserTitleType = {
  title: string;
  browserId: string;
};

type UpdateBrowserFavType = {
  favicon: string;
  browserId: string;
};

type UpdateBrowserCertificateFingerprintType = {
  certificateErrorFingerprint: string | null;
  browserId: string;
};

interface BoardState {
  board: BoardType;
}

const boardId = v4();
const newBoard = {
  id: boardId,
  label: 'New collection',
  browsers: [],
  activeBrowser: null,
  closedUrls: [],
  isFullSize: false,
  browsersActivity: [],
  height: 0,
};

export const initialState: BoardState = {
  board: newBoard,
};

const addCLosedUrl = (state: WritableDraft<BoardState>, url: string) => {
  if (state.board.closedUrls.length > 20) {
    state.board.closedUrls.splice(0, 1);
  }
  state.board.closedUrls.push(url);
};

const addBrowserActivity = (
  state: WritableDraft<BoardState>,
  browserId: string
) => {
  if (state.board.browsersActivity.length > 20) {
    state.board.browsersActivity.splice(0, 1);
  }
  state.board.browsersActivity = state.board.browsersActivity.filter(
    (b) => b !== browserId
  );
  state.board.browsersActivity.push(browserId);
};

const findAndSelectTabForExtensionsFromBrowserId = (
  state: WritableDraft<BoardState>,
  browserId: string
): void => {
  const wcId = state.board.browsers.find(
    (b) => b.id === browserId
  )?.webContentsId;
  if (wcId) window.app.browser.select(wcId);
};

const setStateActiveBrowser = (
  state: WritableDraft<BoardState>,
  browserId: string
): void => {
  state.board.activeBrowser = browserId;
  addBrowserActivity(state, browserId);
  findAndSelectTabForExtensionsFromBrowserId(state, browserId);
};

const getBrowserIndexFromBrowserId = (
  state: WritableDraft<BoardState>,
  browserId: string
): number => {
  return state.board.browsers.findIndex((b) => b.id === browserId);
};

export const boardSlice = createSlice({
  name: 'board',
  initialState,
  reducers: {
    setBoard: (state, action: PayloadAction<BoardType>) => {
      state.board = action.payload;
      const activeBrowser = state.board.browsers.find(
        (b) => b.id === state.board.activeBrowser
      );
      if (!state.board.activeBrowser || !activeBrowser) {
        if (state.board.browsers.length > 0) {
          setStateActiveBrowser(
            state,
            state.board.browsers[state.board.browsers.length - 1].id
          );
        }
      }
      if (state.board.activeBrowser) {
        findAndSelectTabForExtensionsFromBrowserId(
          state,
          state.board.activeBrowser
        );
      }
    },
    toggleBoardFullSize: (state) => {
      state.board.isFullSize = !state.board.isFullSize;
      window.app.analytics.event(`toggle_fullsize_${!state.board.isFullSize}`);
    },
    setActiveBrowser: (state, action: PayloadAction<string>) => {
      setStateActiveBrowser(state, action.payload);
    },
    addBrowser: (state, action: PayloadAction<BrowserProps>) => {
      if (action.payload.browserIndex) {
        state.board.browsers.splice(
          action.payload.browserIndex,
          0,
          action.payload
        );
      } else {
        state.board.browsers.push(action.payload);
      }
      setStateActiveBrowser(state, action.payload.id);
      window.app.analytics.event('add_browser');
    },
    setBrowsers: (state, action: PayloadAction<BrowserProps[]>) => {
      state.board.browsers = action.payload;
    },
    updateBrowser: (state, action: PayloadAction<UpdateBrowserType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex] = {
          ...state.board.browsers[browserIndex],
          ...action.payload.params,
        };
      }
    },
    updateBrowserUrl: (state, action: PayloadAction<UpdateBrowserUrlType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].url = action.payload.url;
      }
    },
    updateBrowserLoading: (
      state,
      action: PayloadAction<UpdateBrowserLoadingType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].isLoading = action.payload.isLoading;
      }
    },
    updateBrowserTitle: (
      state,
      action: PayloadAction<UpdateBrowserTitleType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].title = action.payload.title;
      }
    },
    updateBrowserFav: (state, action: PayloadAction<UpdateBrowserFavType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].favicon = action.payload.favicon;
      }
    },
    renameBoard: (state, action: PayloadAction<string>) => {
      state.board.label = action.payload;
      window.app.analytics.event('rename_board');
    },
    minimizeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);
      state.board.browsers[browserIndex].isMinimized = true;
    },
    unminimizeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);

      state.board.browsers[browserIndex].isMinimized = false;
    },
    removeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);

      // remove browser from state
      if (browserIndex > -1) {
        if (!state.board.closedUrls) state.board.closedUrls = [];
        addCLosedUrl(state, state.board.browsers[browserIndex].url);
        state.board.lastClosedBrowserDimensions = [
          state.board.browsers[browserIndex].width,
          state.board.browsers[browserIndex].height,
        ];
        state.board.browsers.splice(browserIndex, 1);

        // clean browsersActivity state
        state.board.browsersActivity = state.board.browsersActivity.filter(
          (b) => b !== action.payload
        );
      }

      // clean activeBrowser
      if (state.board.activeBrowser === action.payload) {
        if (state.board.browsers.length > 0) {
          setStateActiveBrowser(
            state,
            state.board.browsersActivity[
              state.board.browsersActivity.length - 1
            ]
          );
        } else {
          state.board.activeBrowser = null;
          window.app.browser.selectBrowserView();
        }
      }

      window.app.analytics.event('close_browser');
    },
    removeAllBrowsersExcept: (state, action: PayloadAction<string>) => {
      state.board.browsers
        .filter((b) => b.id !== action.payload)
        .forEach((b) => addCLosedUrl(state, b.url));

      state.board.browsers = state.board.browsers.filter(
        (b) => b.id === action.payload
      );

      // clean activeBrowser
      setStateActiveBrowser(state, action.payload);

      // clean browsersActivity state
      state.board.browsersActivity = state.board.browsersActivity.filter(
        (b) => b === action.payload
      );

      // send event
      window.app.analytics.event('close_others_browser');
    },
    updateBrowserCertificateErrorFingerprint: (
      state,
      action: PayloadAction<UpdateBrowserCertificateFingerprintType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].certificateErrorFingerprint =
          action.payload.certificateErrorFingerprint;
      }
    },
    removeAllBrowsers: (state) => {
      if (!state.board.closedUrls) state.board.closedUrls = [];
      state.board.browsers.forEach((b) => addCLosedUrl(state, b.url));
      state.board.browsers = [];
      state.board.browsersActivity = [];
      state.board.activeBrowser = null;
      window.app.browser.selectBrowserView();
      window.app.analytics.event('close_all_browsers');
    },
    removeLastClosedUrl: (state) => {
      state.board.closedUrls.splice(state.board.closedUrls.length - 1, 1);
    },
    toggleSearch: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);
      state.board.browsers[browserIndex].isSearching =
        !state.board.browsers[browserIndex].isSearching;
    },
    setLastResizedBrowserDimensions: (
      state,
      action: PayloadAction<[number, number]>
    ) => {
      state.board.lastResizedBrowserDimensions = action.payload;
    },
    setBoardHeight: (state, action: PayloadAction<number>) => {
      state.board.height = action.payload;
    },
  },
});

export const {
  setBoard,
  updateBrowserUrl,
  updateBrowserFav,
  renameBoard,
  removeBrowser,
  minimizeBrowser,
  unminimizeBrowser,
  removeAllBrowsers,
  addBrowser,
  setActiveBrowser,
  updateBrowserTitle,
  updateBrowser,
  toggleBoardFullSize,
  removeLastClosedUrl,
  setBrowsers,
  removeAllBrowsersExcept,
  updateBrowserLoading,
  updateBrowserCertificateErrorFingerprint,
  toggleSearch,
  setLastResizedBrowserDimensions,
  setBoardHeight,
} = boardSlice.actions;

export default boardSlice.reducer;
