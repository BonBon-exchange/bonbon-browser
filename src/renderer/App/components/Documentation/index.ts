import { Documentation } from './Documentation';

export * from './Documentation';
export default Documentation;
