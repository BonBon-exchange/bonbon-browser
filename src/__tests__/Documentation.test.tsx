/* eslint-disable react/jsx-props-no-spreading */
import '@testing-library/jest-dom';
import { render, act, fireEvent, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { Middleware } from '@reduxjs/toolkit';

import { Documentation } from '../renderer/App/components/Documentation';
import { mockWindow } from './beforeAll';
import { initialState } from '../renderer/App/store/reducers/Board';

let store: any;
const middlewares: Middleware[] = [];

const props = {
  handleClose: jest.fn(),
};

describe('Documentation', () => {
  beforeAll(() => {
    mockWindow();
    const mockStore = configureStore(middlewares);
    store = mockStore({ board: initialState });
  });

  it('should render', () => {
    expect(
      render(
        <Provider store={store}>
          <Documentation {...props} />
        </Provider>
      )
    ).toBeTruthy();
  });

  it('should show Webpages windows', async () => {
    act(() => {
      render(
        <Provider store={store}>
          <Documentation {...props} />
        </Provider>
      );
    });

    act(() => {
      fireEvent.click(screen.getAllByTestId('documentation-webpages-link')[0]);
    });

    expect(
      screen.getAllByTestId('documentation-webpages-page')[0]
    ).toBeTruthy();
  });
});
