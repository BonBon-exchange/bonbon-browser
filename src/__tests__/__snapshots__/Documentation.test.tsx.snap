// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documentation should render 1`] = `
<div
  id="Documentation__container"
>
  <div
    className="CloseButton__container"
    data-testid="close-button"
    onClick={[MockFunction]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="CloseIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
  <div
    id="Documentation__left-panel"
  >
    <h2>
      Documentation
    </h2>
    <ul>
      <li
        onClick={[Function]}
      >
        Keyboard shortcuts
      </li>
      <li
        data-testid="documentation-webpages-link"
        onClick={[Function]}
      >
        Webpages windows
      </li>
    </ul>
  </div>
  <div
    id="Documentation__right-panel"
  >
    <h2>
      Documentation
    </h2>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + f - Search in a webpage
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL and F at the same time will open a search area.
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + t - Open a new webpage
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL and T at the same time will open a new webpage (named Tab in other browsers).
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + SHIFT + t - Reopen last closed webpage
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL, SHIFT and T at the same time will reopen a last closed webpage (named Tab in other browsers).
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + r - Reload webpage
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL and R at the same time will reload the current focused webpage (named Tab in other browsers).
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + w - Close webpage
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL and W at the same time will close the current focused webpage (named Tab in other browsers). If no webpage are open in the collection (upper tabulation), then the collection will be closed.
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + SHIFT + w - Close collection
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL, SHIFT and W at the same time will close the current collection (upper tabulation).
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + TAB - Focus the next webpage
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL and TAB at the same time will focus the next webpage (named Tab in other browsers).
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        CTRL + SHIFT + TAB - Focus the next collection
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing CTRL, SHIFT and TAB at the same time will focus the next collection (upper tabulation).
      </div>
    </div>
    <div
      className="Documentation__item"
    >
      <div
        className="Documentation__item-title"
      >
        ALT + Scroll - Efficiently scroll in the document
      </div>
      <div
        className="Documentation__item-description"
      >
        Pressing ALT and Scrolling at the same time will scroll into the document without stopping hover the webpages.
      </div>
    </div>
  </div>
</div>
`;
